### PR TITLE
Provenance action parsing

### PR DESCRIPTION
**Pull request by thermokarst**
Right now edges represent actions and nodes represent artifacts. We can tweak this in case the opposite seems more reasonable, but for now I wanted to structure it the same as https://github.com/qiime2/q2view (well, I guess it isn't quite the same, but close enough). For now I see advantages to both options, but this way seems to lend itself well to the "checkpoint recovery" scenario (e.g. "what do I have here?").

## Minimum Working Example

```python
# wget https://docs.qiime2.org/2.0.6/data/tutorials/moving-pictures/taxa-bar-plots.qzv
import pprint

import networkx as nx

from qiime import Visualization
from qiime.core.archive.format.util import parse_actions

pp = pprint.PrettyPrinter(width=100, compact=True)

v = Visualization.load('/path/to/taxa-bar-plots.qzv')
prov = parse_actions(v)

rev_prov = prov.reverse(copy=True)
for to, frm in reversed(list(nx.dfs_edges(rev_prov, v.uuid))):
    print("%s -> %s" % (frm, to))
    pp.pprint(rev_prov[to][frm])
```
### Results:
```
213e1e82-b767-4fb5-b306-f55e58bbcdf2 -> 0b1ce057-9486-4a47-9685-40aa77cb7bc0
{'action': 'classify',
 'parameters': OrderedDict([('chunk_size', 262144), ('n_jobs', 1), ('pre_dispatch', '2*n_jobs'),
                            ('confidence', -1.0)]),
 'plugin': ForwardRef(reference='environment:plugins:feature-classifier'),
 'type': 'method'}
NoProvenance(uuid='b7408a5a-40b8-466a-b1bc-a1b9cb7d7d5e') -> 0b1ce057-9486-4a47-9685-40aa77cb7bc0
{'action': 'classify',
 'parameters': OrderedDict([('chunk_size', 262144), ('n_jobs', 1), ('pre_dispatch', '2*n_jobs'),
                            ('confidence', -1.0)]),
 'plugin': ForwardRef(reference='environment:plugins:feature-classifier'),
 'type': 'method'}
0b1ce057-9486-4a47-9685-40aa77cb7bc0 -> 6833950a-3a64-454e-b3f4-0f1c00593067
{'action': 'barplot',
 'parameters': OrderedDict([('metadata', MetadataPath(path='metadata.tsv'))]),
 'plugin': ForwardRef(reference='environment:plugins:taxa'),
 'type': 'visualizer'}
c6486239-1e75-4d3e-9f53-f010b87577d1 -> 2531ab2d-45f1-4ba1-971a-05cbc9163df2
{'action': 'emp',
 'parameters': OrderedDict([('barcodes', MetadataPath(path='barcodes.tsv')),
                            ('rev_comp_barcodes', False), ('rev_comp_mapping_barcodes', False)]),
 'plugin': ForwardRef(reference='environment:plugins:demux'),
 'type': 'method'}
2531ab2d-45f1-4ba1-971a-05cbc9163df2 -> 15bb3eb0-6167-483e-b92b-687c66a9df13
{'action': 'denoise',
 'parameters': OrderedDict([('trunc_len', 100), ('trim_left', 0), ('max_ee', 2), ('truncq', 2),
                            ('hashed_feature_ids', True)]),
 'plugin': ForwardRef(reference='environment:plugins:dada2'),
 'type': 'method'}
15bb3eb0-6167-483e-b92b-687c66a9df13 -> 6833950a-3a64-454e-b3f4-0f1c00593067
{'action': 'barplot',
 'parameters': OrderedDict([('metadata', MetadataPath(path='metadata.tsv'))]),
 'plugin': ForwardRef(reference='environment:plugins:taxa'),
 'type': 'visualizer'}
```
### Note to reviewer: When merging, please don't squash, both commits represent logical units of work.

### CODE REVIEW

**Comment by ebolyen** [comment]
Not reviewing atm, but just a quick note: the list of dicts representation is really just an artifact of the yaml format, I don't know if we want to propagate that forward. An OrderedDict would be more natural.

File: qiime/core/archive/format/util.py [hunk 1]
```diff
@@ -22,3 +28,64 @@ def artifact_version(version):
...
+    if prov_dir is None or not prov_dir.exists():
+        prov.add_node(artifact.uuid)
+        return prov
+
+    def dict_list_to_ordereddict(l):
+        return OrderedDict([tuple(x.items())[0] for x in l])
+
+    def load_yaml(prov_dir):
+        action_yaml = prov_dir / 'action' / 'action.yaml'
+        metadata_yaml = prov_dir / 'metadata.yaml'
```
Comment (ebolyen) [inline qiime/core/archive/format/util.py]: `ProvenanceCapture` has some of these constants, but maybe they aren't quite as useful where they are? These details strike me as specific to the format version, so some refactoring may be needed in the future.

**Comment by ebolyen** [comment]
I think I am ready to start using this, @thermokarst when you have a chance could you fix the merge conflicts and we'll merge this. I'd like to use this to derive an `ActionGraph` object which I will reuse in the `ScheduledResult` implementation.

**Comment by thermokarst** [comment]
@ebolyen, sounds good! PR has been updated!

**Comment by thermokarst** [comment]
Saving for a rainy day.

**Pull request closed without merging**